### ISSUE TITLE
DOC: The documentation of broadcast_to is ambiguous

### ISSUE THREAD

**Issue by krikru** [numpy/numpy#24767]
### Issue with current documentation:

When reading the documentation page of [numpy.broadcast_to](https://numpy.org/devdocs/reference/generated/numpy.broadcast_to.html#numpy.broadcast_to), is is not clear to me how NumPy determines the locations of the new dimensions when one or several of them have the same sizes (or lengths) as some of the old dimensions. The example provided in the documentation page highlights this ambiguity:

```
>>> x = np.array([1, 2, 3])
>>> np.broadcast_to(x, (3, 3))
array([[1, 2, 3],
       [1, 2, 3],
       [1, 2, 3]])
```

Now, in this example, apparently NumPy decides that it is the first dimension that is the new dimension, while it is the second dimension that should be transferred from `x`. However, I don't see anything on the documentation page that prevents NumPy from making the reversed decision and providing you with this array instead:

```
>>> x = np.array([1, 2, 3])
>>> np.broadcast_to(x, (3, 3))
array([[1, 1, 1],
       [2, 2, 2],
       [3, 3, 3]])
```

So, how does NumPy determine the location of the new dimension in this case? Does it flip a coin?

**Comment by mattip** [comment]
The behaviour of broadcast is described on the [dedicated topic](https://numpy.org/devdocs/user/basics.broadcasting.html#general-broadcasting-rules), where the rules are:
> When operating on two arrays, NumPy compares their shapes element-wise. It starts with the trailing (i.e. rightmost) dimension and works its way left. Two dimensions are compatible when
> - they are equal, or
> - one of them is 1.
> If these conditions are not met, a `ValueError: operands could not be broadcast together exception` is thrown, indicating that the arrays have incompatible shapes. 
>
> Input arrays do not need to have the same number of dimensions. The resulting array will have the same number of dimensions as the input array with the greatest number of dimensions ...

In your example, `x.shape` is `(3,)`, it only has one dimension. Maybe you mean to write `x = np.array([[1, 2, 3]])` which would have shape `(1, 3)`? In both cases, the conditions are met for prepending a 3 to the shape.

Perhaps the broadcasting documentation could be linked to somewhere on the page you point to.

**Comment by krikru** [comment]
@rkern Could we maybe discuss whether the documentation is clear enough before we decide not to fix the issue? The fix could be as easy as linking to the broadcasting documentation as @mattip suggested, along with a comment about how the broadcasting rules tell you how to know the location of the new dimensions (or even simply just that comment, without a link to the broadcasting rules). Or do you think the documentation is fine just because you understand how to use the function?

**Comment by rkern** [comment]
While I think it would be fine to add a link to the broadcasting page, I also think it's fine not to (and I'm marginally against it). In general, we've taken the view that we don't always link to the big, fundamental, pervasive design concepts in every docstring where those concepts are used. Because they are so all-pervasive in the library, that'd be a lot of duplicative links. `np.broadcast_to()` is generally *not* where people first encounter broadcasting (typically, it's binary operations like `+` and `*`); it's a relatively specialized function. People who use it to achieve that specialized effect typically already know the details, or know where it is documented when they need to look up the details.

You happened upon `np.broadcast_to()` early, but in my experience, that's pretty unusual. Sometimes asking and answering the one question is a fine response to a rare occurrence. It doesn't always mean that the documentation needs to be changed. The experience I'm applying here isn't my familiarity with the function, but my long experience with what kinds of questions get asked and reasked.

**Comment by krikru** [comment]
> In your example, `x.shape` is `(3,)`, it only has one dimension. Maybe you mean to write `x = np.array([[1, 2, 3]])` which would have shape `(1, 3)`? In both cases, the conditions are met for prepending a 3 to the shape.

No, I didn't mean to write `x = np.array([[1, 2, 3]])`, since (as I wrote) I copied the example straight from the documentation page. Also, if `x` has the shape `(1, 3)`, then prepending a 3 to the shape would yield an array with the shape `(3, 1, 3)`, which does not match the target shape that was used in the example (which was `(3, 3)`), so I'm not completely sure I understand your comment.

**Comment by rkern** [comment]
He didn't actually mean prepending a `3` in both cases exactly; he meant prepending a `3` in the `(3,)` case and broadcasting the `1` to a `3` in the `(1, 3)` case. In both cases, the broadcasting is happening in an axis before the existing `3` axis. That is, in the resulting `(3, 3)` array, the first axis is broadcasted from the original array.

If you wanted the broadcasting to be along the second axis, then you need a `(3, 1)` array to begin with:

```python
>>> y = x[:, np.newaxis]
>>> y.shape
(3, 1)
>>> y
array([[1],
       [2],
       [3]])
>>> np.broadcast_to(y, (3, 3))
array([[1, 1, 1],
       [2, 2, 2],
       [3, 3, 3]])
```

**Comment by krikru** [comment]
@rkern So do you think it's better to let people who look at the documentation page for `np.broadcast_to` and arrive at the same question as I did go to a discussion forum (or an issue tracker as I apparently did) to ask about how the function works, instead of actually documenting the behavior of the function in the documentation? (I don't know what experience with users you have, but take me as the example to show that users exist who don't have that knowledge 🙂) IMO, the documentation should fully explain what the function does (this could include linking to other pages), and not assume that the user is already partly familiar with the library.

**Comment by rkern** [comment]
Again, I think it'd be fine to add such a link, and if another core dev wants to just go ahead and do it, I'm not going to object to the PR. Honestly, if it weren't for the fact that we're currently being hit with a bunch of students opening low-effort PRs for class credit right now, I'd have been less trigger-happy on closing the issue as an attractive nuisance for them.

I do take this issue as evidence seriously, but it's one data point balanced against over two decades of my personal experience helping newcomers use Numpy and its predecessors. I sympathize that it looms larger for you than it does for me. I appreciate the "link every concept referred to so every docstring is self-contained" documentation philosophy, but it's not the only valid one and not the one that we've chosen. We don't assume that you're perfectly familiar with the whole library to read a single docstring, but we do assume that you've read the [Fundamentals](https://numpy.org/doc/1.26/user/basics.html) or can refer to them when you hit an unfamiliar word. We document those concepts there so we don't have to repeat them everywhere.
